docs/spec.md

```diff
@@ -2331,47 +2341,58 @@ planning to address this in
 
 #### Outputs
 
-| Name     | Type   | Constraints |
-|----------|--------|-------------|
-| `result` | tensor | (C13)       |
+| Name     | Type                       | Constraints                |
+|----------|----------------------------|----------------------------|
+| `result` | tensor or quantized tensor | (C12), (C15), (C17), (C20) |
 
 #### Constraints
 
-* (C1) `lhs` and `rhs` have the same element type.
-* (C2) size(`lhs_batching_dimensions`) $=$ size(`rhs_batching_dimensions`).
-* (C3) size(`lhs_contracting_dimensions`) $=$
+<!-- markdownlint-disable line-length -->
+* (C1) size(`lhs_batching_dimensions`) $=$ size(`rhs_batching_dimensions`).
+* (C2) size(`lhs_contracting_dimensions`) $=$
   size(`rhs_contracting_dimensions`).
-* (C4) `lhs_batching_dimensions` and `lhs_contracting_dimensions` combined are
+* (C3) `lhs_batching_dimensions` and `lhs_contracting_dimensions` combined are
   unique.
-* (C5) `rhs_batching_dimensions` and `rhs_contracting_dimensions` combined are
+* (C4) `rhs_batching_dimensions` and `rhs_contracting_dimensions` combined are
   unique.
-* (C6) 0 $\le$ `lhs_batching_dimensions[i]` $\lt$ rank(`lhs`) for all `i`
+* (C5) 0 $\le$ `lhs_batching_dimensions[i]` $\lt$ rank(`lhs`) for all `i`
   $\in$ [0, size(`lhs_batching_dimensions`)).
-* (C7) 0 $\le$ `lhs_contracting_dimensions[i]` $\lt$ rank(`lhs`) for all `i`
+* (C6) 0 $\le$ `lhs_contracting_dimensions[i]` $\lt$ rank(`lhs`) for all `i`
   $\in$ [0, size(`lhs_contracting_dimensions`)).
-* (C8) 0 $\le$ `rhs_batching_dimensions[d]` $\lt$ rank(`rhs`) for all `i`
+* (C7) 0 $\le$ `rhs_batching_dimensions[i]` $\lt$ rank(`rhs`) for all `i`
   $\in$ [0, size(`rhs_batching_dimensions`)).
-* (C9) 0 $\le$ `rhs_contracting_dimensions[d]` $\lt$ rank(`rhs`) for all `i`
+* (C8) 0 $\le$ `rhs_contracting_dimensions[i]` $\lt$ rank(`rhs`) for all `i`
   $\in$ [0, size(`rhs_contracting_dimensions`)).
-* (C10) dim(`lhs`, `lhs_batching_dimensions[i]`) $=$
+* (C9) dim(`lhs`, `lhs_batching_dimensions[i]`) $=$
   dim(`rhs`, `rhs_batching_dimensions[i]`) for all `i` $\in$ [0,
   size(`lhs_batching_dimensions`)).
-* (C11) dim(`lhs`, `lhs_contracting_dimensions[i]`) $=$
+* (C10) dim(`lhs`, `lhs_contracting_dimensions[i]`) $=$
   dim(`rhs`, `rhs_contracting_dimensions[i]`) for all `i` $\in$ [0,
   size(`lhs_contracting_dimensions`)).
-* (C12) size(`precision_config`) $=$ 2.
-* (C13) shape(`result`) $=$ dim(`lhs`, `lhs_batching_dimensions`) +
+* (C11) size(`precision_config`) $=$ 2.
+* (C12) shape(`result`) $=$ dim(`lhs`, `lhs_batching_dimensions`) +
   dim(`lhs`, `lhs_result_dimensions`) + dim(`rhs`, `rhs_result_dimensions`).
+* If the operation uses non-quantized tensors:
+  * (C13) element_type(`lhs`) $=$ element_type(`rhs`).
+* If the operation uses quantized tensors:
+  * (C14) `is_quantized_tensor(lhs) and is_quantized_tensor(rhs)`.
+  * (C15) `storage_type(lhs) =  storage_type(rhs) = storage_type(result)`.
```

Comment:
I think we may want to allow `storage_type(result) != storage_type(rhs)` so that it is possible to dump the summation directly before applying the activation function.

For example:

```
%t = "stablehlo.dot_general"(%input, %param) : (tensor<... x !quant.uniform<ui8:f32, s1:zp1>>, tensor<... x !quant.uniform<ui8:f32, s2:zp2>> -> tensor<... x !quant.uniform<i32:f32, s3:zp3>>
%dq = "stablehlo.dequantize"(%t) : (tensor<... x !quant.uniform<i32:f32, s3:zp3>>) -> tensor<... x f32>
%actfn = "stablehlo.logistic"(%dq) : (tensor<... x f32>) -> tensor<... x f32>
%q = "stablehlo.quantize"(%actfn) : (tensor<... x f32>) -> ...
```

Reply:
Thanks for pointing that!I agree with the point. 
Additionally, even for non-quantized type, we do allow `element_type(result)` is be different from `element_type(lhs) (or element_type(rhs))`. 

Also, I note that as per https://github.com/openxla/stablehlo/issues/1439, we will be revisiting the type constraint between `lhs` and `rhs`. 